### PR TITLE
Add support for project Java version update based on extensions

### PR DESCRIPTION
**Pull request by ia3andy**
Java version is now updated based on the extension metadata (min java version).

Gradle is not yet updating as we are still using OpenRewrite 7.x

part of #32392

### CODE REVIEW

**Comment by ia3andy** [comment]
cc @maxandersen @gsmet

**Comment by ia3andy** [comment]
Now we should decide:
- should we first merge this
- is it not to risky to put it in the next release
- what should we start first?
  - fetch external extension recipes and add them
  - auto-add recipes to update non-platform extension versions
  - java version update

**Comment by maxandersen** [comment]
> Non platform extensions version update auto generated

fixing this means we can removing  https://github.com/quarkusio/quarkus-updates/blob/main/recipes/src/main/resources/quarkus-updates/core/3alpha.yaml#L30, right?

> Fetch recipes from quarkus-updates for the extensions updates

This will be nice, but in theory its not an absolute blocker as long as the rules needed by i.e. camel can be added to the quarkus-updates. If we know of migrations that would be problematic that would change.

> Update Java version based on min Java version

This is  a minimal requirement for camel as they must update to java 17.

**Comment by ia3andy** [comment]
> > Non platform extensions version update auto generated
> 
> fixing this means we can removing https://github.com/quarkusio/quarkus-updates/blob/main/recipes/src/main/resources/quarkus-updates/core/3alpha.yaml#L30, right?

Yep

> 
> > Fetch recipes from quarkus-updates for the extensions updates
> 
> This will be nice, but in theory its not an absolute blocker as long as the rules needed by i.e. camel can be added to the quarkus-updates. If we know of migrations that would be problematic that would change.
> 
> > Update Java version based on min Java version
> 
> This is a minimal requirement for camel as they must update to java 17.

Ok, I've set the order in the TODO list

**Comment by ia3andy** [comment]
I added a new Commit to:
- Clean the code related to Java version
- Detect the version on Gradle and Maven project
- Add the JavaVersion to the QuarkusProject
- Warn if the project JavaVersion is not compatible with one or more of the udpated extensions

I haven't tested Gradle at all.

I've tested a Maven project in Quarkus 2.16 with Java 11 and the camel-core extension but currently the managed extensions are not in the `ProjectExtensionsUpdateInfo` so it requires some changes to include it in there so that `extensionsUpdateInfo.getMinJavaVersion()` also includes those extensions.

I've checked the data is here, we just need to put it in the UpdateInfo:
image

It might be also nice to display that they are going to be updated by the platform bom.

**Comment by ia3andy** [comment]
On Gradle we can use this: https://docs.openrewrite.org/recipes/gradle/changejavacompatibility
And on Maven we can update the properties, as `maven.compiler.release` is the one we use on generated project I would say we only set that one.

**Comment by ia3andy** [comment]
The Gradle recipe is not applied because it's meant for OR 8.x, good news is that it doesn't break and the update java version log is displayed

**Comment by ia3andy** [comment]
@gsmet CI passed or not?

**Comment by ia3andy** [comment]
@gsmet up

**Comment by gsmet** [comment]
Sorry, I need to have a look at that.

**Review by gsmet — COMMENTED** [review]
I tested this PR with and without a Camel extension and it seems to work OK.

I added a small stylistic comment but I think we should probably get this in and also backport it to 3.2.

**Review by aloubyansky — APPROVED** [review]

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/JavaVersion.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public String toString() {
+        return version;
+    }
+
+    // ordering is important here, so let's keep them ordered
+    public static final SortedSet<Integer> JAVA_VERSIONS_LTS = new TreeSet<>(List.of(11, 17));
+    public static final int DEFAULT_JAVA_VERSION = 11;
+    public static final int MAX_LTS_SUPPORTED_BY_KOTLIN = 17;
+    public static final String DETECT_JAVA_RUNTIME_VERSION = "<<detect java runtime version>>";
+    public static final Pattern JAVA_VERSION_PATTERN = Pattern.compile("(\\d+)(?:\\..*)?");
```
[the remark this answers is not in the document]
Comment (gsmet) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/JavaVersion.java:61]: @ia3andy  I will merge this for now but I would appreciate if you could give it some love when you're back. Thanks!

**Review by gsmet — APPROVED** [review]

**Pull request merged**